### PR TITLE
feat(analyze): Add DCM and fix potentiall bugs, depreciations 

### PR DESCRIPTION
**Pull request by mhadaily**
## Description

This pull request introduces several changes, primarily focusing on integrating Dart Code Metrics (DCM) into the CI workflow, updating the contributing guidelines, and making various code improvements and refactorings. The most important changes are summarized below:

### CI/CD Workflow Enhancements:
* Added a step to install DCM in the GitHub Actions workflow (`.github/workflows/checks.yml`).
* Updated the analysis step to use DCM for both Flutter and Dart in the GitHub Actions workflow (`.github/workflows/checks.yml`).

### Documentation Updates:
* Updated `CONTRIBUTING.md` to include steps for installing and activating DCM, and clarified the use of DCM in the contribution process.

### Code Quality and Linting:
* Added Dart Code Metrics configuration to `analysis_options.yaml` and other related files to enforce specific linting rules. [[1]](diffhunk://#diff-9a967a7bb0393381dfdf9cce102a9cea39146829ab8b38e983d54ac8827486bcR2-R26) [[2]](diffhunk://#diff-d0d0c3c48e12e7f7de2ecf83e8b94c71d22d98de6e5a4c9378eebf61fc6b8115R2-R7) [[3]](diffhunk://#diff-cf4545906979572dba7ee9b61debc7b1e05089aee5b30d4c314932b95eab982bL11-R19)

### Code Refactoring:
* Refactored multiple files to improve code readability and maintainability, including changes to widget properties and method calls. [[1]](diffhunk://#diff-292c5a714b4e4470af917c86cc6f4c1a2effad4c48d3353a591761fd648fed13L22-R22) [[2]](diffhunk://#diff-808d55dbfda3b0c77dffd539a6b03a17c897d719329dc2d295a03820bc871c40L15-R15) [[3]](diffhunk://#diff-34f0c9b96d50aea1eacbcc1faf9efbb3d29dec325400cd4b121a946ab058ab60L5-L8) [[4]](diffhunk://#diff-474cf512e7602997cf6f9c439fd394341d2b26f5b1d0daf90817d427bc4e6033L5-R9) [[5]](diffhunk://#diff-7b63bd18616401dbc0a40ed257022401827db35be82312eb74b98fac13bb603cL30-R46) [[6]](diffhunk://#diff-4d8c3e0b3c872c5505fdc4b9cf1f8ce9ca4bdd176a76868df02646b5bf136990L77-R77) [[7]](diffhunk://#diff-d4fe755db4210a58e7f2e920759170d82026f3772f4f6b4d6397c3ac26af3271L49-R49)

### Minor Code Improvements:
* Made minor adjustments to formatting and method calls to enhance code consistency and readability. [[1]](diffhunk://#diff-ba8d96e1953662ca8daa138fb13ca79e7a503a2b8b0a0ec80a3ae6cfa99bcb6fL59-R61) [[2]](diffhunk://#diff-c443ee947abc9c4fbd27d013dd6e3a55a99a83138c17b0f45b1efdcf864259ccL202-R202) [[3]](diffhunk://#diff-693f7972fbcdec4ed56a9235d9644f34b9d7c30bfbadf150394ad728f1f592d6L4-R16) [[4]](diffhunk://#diff-92fc907c205ba20bd2480f32895e7775a0e95a009039e5f99a8f11e66e9239e7L3)

These changes collectively improve the project's CI/CD pipeline, documentation, code quality, and overall maintainability.

## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database](https://github.com/woltapp/wolt_modal_sheet/issues). Indicate, which of these issues are resolved or fixed by this PR.*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors.
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] The package compiles with the minimum Flutter version stated in the [pubspec.yaml](https://github.com/woltapp/wolt_modal_sheet/blob/main/pubspec.yaml#L8)

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[Contributor Guide]: https://github.com/woltapp/wolt_modal_sheet/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by ulusoyca** [comment]
@mhadaily Can we have an example here to demonstrate how to create an internal static analysis package that includes all the rules and is also added as dependency to the demo apps and also to the package lib? I see that currently every demo app has its own analysis option. We can add baseline to silence warnings in the demo apps if they are too big changes.

File: lib/src/content/components/main_content/wolt_modal_sheet_top_bar_title.dart [hunk 1]
```diff
@@ -37,12 +37,12 @@ class _WoltModalSheetTopBarTitleState extends State<WoltModalSheetTopBarTitle> {
   }
 
   void _extractPageTitleText() {
-    final isPageTitleTextWidget = widget.page.pageTitle is Text;
+    // final isPageTitleTextWidget = widget.page.pageTitle is Text;
```
Comment (ulusoyca) [inline lib/src/content/components/main_content/wolt_modal_sheet_top_bar_title.dart:40]: Are we sure this is unnecessary? This is checking if the current widget is text widget. If not, it is checking all the descendents of the current widget.

**Review by ulusoyca — COMMENTED** [review]
We can accept all the suggsted changes in the demo app. My comments are focusing only the package lib. I left minor comments. Huge thanks 🚀

**Comment by ulusoyca** [comment]
@mhadaily Looks like the workflow is expecting an approval to run. Is this intentional? If yes, it does not sound useful since we expect the static analysis feedback from CI to work even in draft PRs.

image

**Comment by incendial** [comment]
> @mhadaily Looks like the workflow is expecting an approval to run. Is this intentional? If yes, it does not sound useful since we expect the static analysis feedback from CI to work even in draft PRs.

Hmm, if that's related to this PR, then that's probably because of how GitHub actions work. Either DCM needs to be installed without the action or another trigger should be used (example: https://github.com/leoafarias/fvm/blob/main/.github/workflows/test.yml#L45).

I'd say the first approach is safer.

File: lib/src/content/components/main_content/wolt_modal_sheet_top_bar_title.dart (hunk 1 above)
Comment (mhadaily) [reply lib/src/content/components/main_content/wolt_modal_sheet_top_bar_title.dart]: revert it, there is a if block that is empty it seems

**Comment by mhadaily** [comment]
> @mhadaily Looks like the workflow is expecting an approval to run. Is this intentional? If yes, it does not sound useful since we expect the static analysis feedback from CI to work even in draft PRs.
> 
> image

@ulusoyca according to this https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-workflow-runs/approving-workflow-runs-from-public-forks 
It seems since I am first time contributor and I am touching some pull_request_target, you need to approve this change and run. 

it looks like this is only for first time contributor and will not happen later. Let's give it a try.

**Comment by mhadaily** [comment]
> We can accept all the suggsted changes in the demo app. My comments are focusing only the package lib. I left minor comments. Huge thanks 🚀

Awesome, first two comments.

**Review by ulusoyca — APPROVED** [review]

**Comment by ulusoyca** [comment]
@mhadaily `WidgetStatePropertyAll ` fill fail because the min Flutter SDK version is 3.13.0+

**Comment by mhadaily** [comment]
> @mhadaily `WidgetStatePropertyAll ` fill fail because the min Flutter SDK version is 3.13.0+

AH that's right, I didn't check that. Should I change min SDK or you prefer I revert the change for "WidgetStatePropertyAll" ?

**Comment by ulusoyca** [comment]
> > @mhadaily `WidgetStatePropertyAll ` fill fail because the min Flutter SDK version is 3.13.0+
> 
> AH that's right, I didn't check that. Should I change min SDK or you prefer I revert the change for "WidgetStatePropertyAll" ?

Actually it is a good idea to change the min SDK in the example apps, without changing the package min SDK improvements.

**Comment by mhadaily** [comment]
Done @ulusoyca

**Review by ulusoyca — APPROVED** [review]
🚢

**Comment by mhadaily** [comment]
btw, @ulusoyca when you got chance, try to enable the "disabled rule" and fix the issue, I find several potential bugs and also memory leaks problems. 

also if you run `dcm check-unused-files` and `dcm check-unused-files` you will see you also have dead code and files. 

generally a good idea to play around with some of the commands

**Comment by ulusoyca** [comment]
@mhadaily Thanks for the setup, I will delete / simplify some demo projects, and add more rules. Thanks for the setup!

**Comment by mhadaily** [comment]
@ulusoyca It seems the main library actually had 3.13 but the examples were not. so I changed everything, Not sure why it needs approval again. it was working without approval but suddenly last commit needs it again.

**Comment by ulusoyca** [comment]
🚢

**Comment by mhadaily** [comment]
It's still failing, I am not sure why?  do you have any idea? 
I checked it on my machine and it seems it's all good 
Screenshot 2024-08-31 at 20 47 11

**Comment by ulusoyca** [comment]
> It's still failing, I am not sure why? do you have any idea? I checked it on my machine and it seems it's all good Screenshot 2024-08-31 at 20 47 11

Your Local is probably the latest flutter but the action is run on 3.13.0. I think the examples should use `MaterialState` instead of `WidgetState`

**Comment by mhadaily** [comment]
Well, that is correct, I thought WidgetState is available after 3.13, anyway I have revert it to MaterialState. Maybe approve again and let's see how it works. 

if it passes then we are good. I can probably deal with depreciation when you are ready.

**Comment by mhadaily** [comment]
@ulusoyca sorry again 😄 

one more time approval. I think this is the final time. Though tests are failing and I think you know that already

**Comment by ulusoyca** [comment]
@mhadaily we are good now!

**Comment by ulusoyca** [comment]
@mhadaily can we merge now?

**Comment by mhadaily** [comment]
Yes please



*Best regards, *
Majid

**Pull request merged**